golang/cosmos/x/swingset/types/msgs.go

```diff
@@ -330,13 +331,86 @@ func (msg MsgInstallBundle) ValidateBasic() error {
 	if msg.Submitter.Empty() {
 		return sdkerrors.Wrap(sdkerrors.ErrInvalidAddress, "Submitter address cannot be empty")
 	}
-	if len(msg.Bundle) == 0 {
+	if len(msg.Bundle) == 0 && len(msg.CompressedBundle) == 0 {
 		return sdkerrors.Wrap(sdkerrors.ErrUnknownRequest, "Bundle cannot be empty")
 	}
+	if len(msg.Bundle) != 0 && len(msg.CompressedBundle) != 0 {
+		return sdkerrors.Wrap(sdkerrors.ErrUnknownRequest, "Cannot submit both a compressed and an uncompressed bundle at the same time")
+	}
+	if len(msg.Bundle) > 0 && msg.UncompressedSize != 0 {
+		return sdkerrors.Wrap(sdkerrors.ErrUnknownRequest, "Uncompressed size cannot be set without a compressed bundle")
+	}
+	if len(msg.CompressedBundle) > 0 && !(msg.UncompressedSize > 0) {
+		return sdkerrors.Wrap(sdkerrors.ErrUnknownRequest, "Uncompressed size must be positive")
+	}
+	if msg.UncompressedSize == math.MaxInt64 {
+		// must avoid the limit because we compute its successor in Uncompress()
```

Comment:
I still think we probably want a far lower limit on the order of tens of megabytes, although gzip itself tops out [a little over 1000x](https://stackoverflow.com/a/16794960) so a 1 MB message limit constrains decompression to about 1 GB, which may be tolerable risk.

Reply:
A 1GB uncompressed size at $2/KB is $2M. Potential griefers would need an account balance which could cover the fees before anyone would attempt decompression. Nevertheless this sounds like a question for @rowgraus.

Reply:
Rowland delegated to Ari who suggested a 10MB limit. Will hardcode that for now. We can make it a governance parameter later if needed - might be obsoleted by chunked updates or smaller per-module bundles.